nx/lib/nx/shape.ex

```diff
@@ -1218,6 +1218,57 @@ defmodule Nx.Shape do
     {List.to_tuple(result_shape), result_names}
   end
 
+  @doc """
+  Returns shape if valid and raises error if not.
+  """
+  def take_diagonal(tensor) do
+    case tensor.shape do
+      {len, breadth} ->
+        {len, breadth}
+
+      _bad_shape ->
+        raise ArgumentError, "take_diagonal/2 expects tensor of rank 2, got:\n#{inspect(tensor)}"
+    end
+  end
+
+  @doc """
+  Validates an offset for given shape
+
+  ## Examples
+
+      iex> Nx.Shape.validate_offset!({3, 4}, 1)
+      :ok
+
+      iex> Nx.Shape.validate_offset!({3, 4}, -1)
+      :ok
+
+  ## Error cases
+
+    iex> Nx.Shape.validate_offset!({3, 4}, 3)
+    ** (ArgumentError) offset must be less than length when positive, got: 3
+
+    iex> Nx.Shape.validate_offset!({3, 4}, -4)
+    ** (ArgumentError) absolute value of offset must be less than breadth when negative, got: -4
+  """
+  def validate_offset!(shape, offset) do
+    {len, breadth} = shape
```

Comment:
This should be matched in the function head to avoid raising a RuntimeError upon invalid shape

Reply:
Okay. I am assuming that the shape is valid, but a check won't hurt.